storage/storage_image.go

```diff
@@ -482,55 +484,66 @@ func (s *storageImageDestination) PutBlob(ctx context.Context, stream io.Reader,
 		Digest: "",
 		Size:   -1,
 	}
-	// Set up to digest the blob and count its size while saving it to a file.
-	hasher := digest.Canonical.Digester()
-	if blobinfo.Digest.Validate() == nil {
-		if a := blobinfo.Digest.Algorithm(); a.Available() {
-			hasher = a.Digester()
+	if blobinfo.Digest != "" {
+		if err := blobinfo.Digest.Validate(); err != nil {
+			return errorBlobInfo, fmt.Errorf("invalid digest %#v: %w", blobinfo.Digest.String(), err)
```

Comment:
FWIW this introduces the first use of Go 1.13-style standard-library errors with `Unwrap()` instead of `github.com/pkg/errors.Wrapf`. We require Go 1.13 anyway, and pkg/errors includes call stack information which we never used, so moving closer to the standard library will hopefully be more efficient as well as, eventually, more familiar to newcomers.

I don’t think we want to do any kind of large-scale migration, just maybe start using this in new code.

@vrothberg RFC.